### ISSUE TITLE
"New Eloquent Model" UI discussion

### ISSUE THREAD

**Issue by adelf** [laravel-idea/plugin#741]
Planned features:

- [x] completions for field names based on model names, like "user_id", "post_id"
- [x] allow adding belongsTo relations
- [x] "Settings" dialog with all post-generation parameters like class name patterns("create_\${MODEL_NAME|className|snakeCase|plural}_table") or should the "authorize" method be generated for form requests
- [x] More post-generations: seeder, Nova/Filament resources, etc.
- [x] Option to generate "$fillable" array in the model class (available in 7.0.1)
- [x] tinyInt and boolean field types (available in 7.0.1)
- [x] add HasFactory trait if there is a factory class (available in 7.0.1)

**Comment by logic-b-o-m-b** [comment]
I would like to see:
- option to select or type in schema used by migration;
- option to use uuids, also for PKs;
- option to include in migration FKs;
- option to include in migration indexes for FKs (postgres);
- option to create composite indexes, with drag and drop for easy column re-ordering;
- option to add comments to fields and indexes ;
- when creating  indexes it would be useful to have features not supported by laravel migrations but supported by postgres (ordering of values, ordering of nulls etc), generation of RAW SQL may be required;
- perhaps all sophisticated options should be visible only after selecting database vendor?
- at some stage feature that allows to generate migration(s) from existing schema, including diff (migrating changes including in the schema, but lacking in existing migrations).

I really like changes in recent version. Thank you!

**Comment by DevRaeph** [comment]
i would love to see:
- default value attribute for migration row
- option like tinyInt or bool

**Comment by silas229** [comment]
Please add "$fillable" checkbox next to nullable which is automatically checked

**Comment by Antons-S** [comment]
Maybe if you check 'Create factory' then model should have `HasFactory` trait
```
use Illuminate\Database\Eloquent\Factories\HasFactory;

class ModelName extends Model
{
    use HasFactory
```

**Comment by adelf** [comment]
@Antons-S totally forgot. Fixed. Thank you!

**Comment by Kriso1337** [comment]
it would be great to see there editable length for string, also decimal type for MySql is also great feature, with editable "total" and "places" fields

**Comment by sneycampos** [comment]
> 

Since we can create Models in non default namespace, e.g: Modules\Buildings\Models, would be interesting if the Factory was created on same namespace (Modules\Buildings\Factories) and uses the newFactory() method on Model to assign the factory correctly, since it is not on default namespace Database\Factories.

**Comment by raveren** [comment]
Great, but it's not using the overriden stubs :/
![image](https://user-images.githubusercontent.com/364220/227194018-d5c10c96-7c2f-4570-adaf-a971c0387b85.png)

**Comment by benounnas** [comment]
when it comes to foreign keys and the suggestions above, i didn't see the mention of the option `cascadeOnDelete` & `cascadeOnUpdate`, Events & Observers, may be a trait attaching area too ?

**Comment by osbre** [comment]
Ideas:
- Ability to set a custom primary key (`$primaryKey`), 
- Ability to set a custom route key (`getRouteKeyName(): string`).

**Comment by bokoch** [comment]
Would be great to have an option to generate php-doc with model attributes.

**Comment by mwaqar666** [comment]
It would be nice to have `$table` property also. It would be optional _(which is the default behavior)_. But you can add a custom table name in `$table` property to associate the model to the provided table name.

Thank you for the great plugin ❤️

**Comment by henryavila** [comment]
This plugin is wonderfull!!

It would be nice to to add the phpdoc `@property`  for every db column on model created.

**Comment by sneycampos** [comment]
> 

Hmm, i don't think so. The plugin itself does this work to avoid these annotations on code...

**Comment by 7met** [comment]
I like to use ULIDs (or UUID) as primary keys, it would be nice to se that as an option, or an overall config for a given site with the HasULIDs or HasUUIDs trait auto added to the model.

Overall though this is really great and love what you've done. Well worth the money.

**Comment by jackwh** [comment]
This is a great addition, thank you! The only thing I'd love to see is auto-generated TypeScript type definitions for our models... 😎😍

A checkbox in this new UI would be great, but it'd be an awesome utility elsewhere in the plugin too. Cheers!

**Comment by himak** [comment]
Please add functions as remove field or setting order fields with moving up and down.

**Comment by bitsmyth** [comment]
Polymorphic checkbox would be nice

**Comment by mcxzyang** [comment]
Please add decimal to the type list and add comment to the field.

**Comment by cornelRaiu** [comment]
An option for generating Pivot models would be nice too.

`php artisan make:model Member --pivot`

**Comment by verkit** [comment]
It would be nice if added button to remove field

**Comment by PatrickJoannisse** [comment]
Just want to second @raveren's point that not leveraging overridden stubs. 

A good example of this is multitenancy and having overridden stubs to make sure the tenant id is added to every model, factory and migration moving forward. This way removing the field becomes a conscious decision rather than an oversight.

**Comment by davidseverwright** [comment]
It would be nice if there was support for enum's too.
Ideally when enum is selected it would have a list of all backed enums to select from, and then automatically pull the values into the migration and add the cast to the model.

Being able to create a new enum as part of creating the model would be extra-nice.

**Comment by adelf** [comment]
> Just want to second @raveren's point that not leveraging overridden stubs.
> A good example of this is multitenancy and having overridden stubs to make sure the tenant id is added to every model, factory and migration moving forward. This way removing the field becomes a conscious decision rather than an oversight.

You(and every package creator for their packages) will be able to add an option to the right part(id, timestamps). If it checked - some fields will be added to migration, some traits to the model... and so on.

**Comment by adelf** [comment]
@davidseverwright good idea. Thanks. I'll try.

**Comment by raveren** [comment]
@adelf there's much more power in the custom stubs and Laravel did not make them customizeable for no reason.

For example I use them to support our HUGE systems historically, unfortunately, custom-named created_at columns and some other stuff. I had to extend `Blueprints` inbuilt class for that.

**Comment by argolian** [comment]
Wonderful plugin which I use a lot. Like others would like to see additional fields for String Sizes. Add Foreign and ForeignIDs, Enum With Options Field (Not 100% sure about this one since there's debate whether we should use enums and the practibility of adding a field list).

**Comment by ghost** [comment]
Would love to be able to reorder fields once I've filled them in, and the ability to choose the `date` type instead of just `datetime` would be useful.
Also the ability to access the rest of the project while the modal is open would be nice, as I often need to refer back to the code while creating a new model to check the relationships.

**Comment by bmckay959** [comment]
Great plugin!

I deal with a bunch of 3rd party data where the table structure is essentially already done. A lot of these have a bunch of columns(50+ sometimes).

- It would be nice when 'tabbing' to create a new field, if the scroll area would auto scroll down to see the new input.

**Comment by KSneijders** [comment]
@bmckay959 It already should ask for confirmation when pressing enter or clicking the `OK` button.
It should be enabled by default, but in case it is not:

![image](https://github.com/laravel-idea/plugin/assets/32707500/2e5f1aec-90c4-4681-88ef-cc2155de5c87)

**Comment by bmckay959** [comment]
Interesting. I don't have that settings button available on MacOS. I also clicked around in the settings area and could not find.

<img width="823" alt="Screenshot 2023-08-25 at 10 26 55 AM" src="https://github.com/laravel-idea/plugin/assets/7681836/a46f8cdd-6963-45b4-9da1-fd49a3873d06">

**Comment by KSneijders** [comment]
> @bmckay959 Interesting. I don't have that settings button available on MacOS. I also clicked around in the settings area and could not find.

Just making sure, your plugin is version `7.2.0.232` and PhpStorm `2023.2`?

**Comment by bmckay959** [comment]
@KSneijders You are right. I was behind on updates. I'll go back and delete my messages to not clog up this threads. Thanks for pointing that out.

**Comment by FoksVHox** [comment]
It would be nice to see the following things:

- The `date` column type, currently it's just the `dateTime` column type.
- The ability to change the table name.
- The ability to limit the length on fields, like `string("my_field", 4)`.
- The ability to mark a field `unique`.
- The ability to add comments on columns.

**Comment by silas229** [comment]
I am not the maintainer or even a contributor to the project, but I think it should be clear to everyone that it is impossible to integrate the entire Eloquent Framework into a single window. Therefore, we have to focus on core elements that are annoying to write by hand. Many features mentioned here are certainly practical, but impossible to implement in their entirety.

**Comment by FoksVHox** [comment]
> I am not the maintainer or even a contributor to the project, but I think it should be clear to everyone that it is impossible to integrate the entire Eloquent Framework into a single window. Therefore, we have to focus on core elements that are annoying to write by hand. Many features mentioned here are certainly practical, but impossible to implement in their entirety.

@silas229 The field types isn't impossible, and so are many of the suggestions.

**Comment by silas229** [comment]
> @silas229 The field types isn't impossible, and so are many of the suggestions.

Of course, it is possible to implement them, but if you want to implement all Eloquent features, the software is no longer usable

**Comment by FoksVHox** [comment]
> > @silas229 The field types isn't impossible, and so are many of the suggestions.
> 
> Of course, it is possible to implement them, but if you want to implement all Eloquent features, the software is no longer usable

I'd disagree, as long as it's designed properly.

**Comment by JurgenThijs** [comment]
I always use the $guarded = ['id'] instead of the $fillable array

**Comment by balanso** [comment]
when i add boolean field in "New eloquent model" i want to see 'boolean' validator in Form request, but i don't see it.

**Comment by lloricode** [comment]
Hi please add search on selecting type, it just convenient if does, thank you

**Comment by Heyian** [comment]
I just want to second the need for customised stubs, it would be great to use this plugin with the custom stubs.

**Comment by Bark-fa** [comment]
It would be great if we could specify default values when creating models

**Comment by Phelms215** [comment]
After seeing the new form, what would be really amazing is a blueprint/visio style linker to create relationships

**Comment by Muetze42** [comment]
`add $fillable array to model class` doesn't work with version 8.0.2.231 in my case.

The option `add $casts array to model class` would be a usefull option.

**Comment by adelf** [comment]
@Muetze42 ID and updated/created fields aren't included in "fillable", so it seems it is empty in your case. Laravel Idea generates the "casts" property if it needs to cast some fields.

**Comment by willem-v-dam** [comment]
> Please add functions as remove field or setting order fields with moving up and down.

If you leave a field empty, it will not be generated.

**Comment by MYHINDA** [comment]
> > I am not the maintainer or even a contributor to the project, but I think it should be clear to everyone that it is impossible to integrate the entire Eloquent Framework into a single window. Therefore, we have to focus on core elements that are annoying to write by hand. Many features mentioned here are certainly practical, but impossible to implement in their entirety.
> 
> @silas229 The field types isn't impossible, and so are many of the suggestions.

right. 
setting order fields would be helpful.
option to customize path of controllers - if i want to save some controllers in a specific sub directory (Such API)
option to change table name with all references.

thank you

**Comment by ricocrivelli** [comment]
> It would be nice to see the following things:
> 
>     * The `date` column type, currently it's just the `dateTime` column type.
> 
>     * The ability to change the table name.
> 
>     * The ability to limit the length on fields, like `string("my_field", 4)`.
> 
>     * The ability to mark a field `unique`.
> 
>     * The ability to add comments on columns.

I'm missing the `date` column type too!

**Comment by sneycampos** [comment]
> > It would be nice to see the following things:
> > ```
> > * The `date` column type, currently it's just the `dateTime` column type.
> > 
> > * The ability to change the table name.
> > 
> > * The ability to limit the length on fields, like `string("my_field", 4)`.
> > 
> > * The ability to mark a field `unique`.
> > 
> > * The ability to add comments on columns.
> > ```
> 
> I'm missing the `date` column type too!

What version are you using? i see "date" column
![image](https://github.com/laravel-idea/plugin/assets/585263/0f31abeb-6ff4-425e-859e-384e02d84e5e)

**Comment by ricocrivelli** [comment]
> > > It would be nice to see the following things:
> > > ```
> > > * The `date` column type, currently it's just the `dateTime` column type.
> > > 
> > > * The ability to change the table name.
> > > 
> > > * The ability to limit the length on fields, like `string("my_field", 4)`.
> > > 
> > > * The ability to mark a field `unique`.
> > > 
> > > * The ability to add comments on columns.
> > > ```
> > 
> > 
> > I'm missing the `date` column type too!
> 
> What version are you using? i see "date" column ![image](https://private-user-images.githubusercontent.com/585263/308559576-0f31abeb-6ff4-425e-859e-384e02d84e5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMjYyODIsIm5iZiI6MTcwOTEyNTk4MiwicGF0aCI6Ii81ODUyNjMvMzA4NTU5NTc2LTBmMzFhYmViLTZmZjQtNDI1ZS04NTllLTM4NGUwMmQ4NGU1ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyOFQxMzEzMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jOGRjZDBlOTM2ODQ0ODg0ZTM5MWE1OGUzOThmZjIzZDU1MmE1M2M1MmFjNzNlMjk5NzI3ZGU5ZDliZGNlNTU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.fD77OBMPr6mmsXyGJYe_57jgycGt4WHg2aP-367wEgo)

Oh! My bad! I was using an old version! Thanks very much @sneycampos ! Brazilians saving brazilians! =)

**Comment by arm092** [comment]
Nice to have:

* Index select-box, default `none`, with values of index types (primary, unique, fulltext...)
* Nullable checkbox for relation fields too
* And may be `Fillable` checkbox

Thank You!

**Comment by altayevrim** [comment]
default value should be great for boolean fields for example (or for every field) because when I create a model with API and test it, omitting boolean fields throws an SQL error.  

Also when I use Jetstream, I have to uncheck Register in Auth Service Provider option in Policy Settings otherwise it gives an error (Register Policy Error - There is no Auth Service Provider classes in the project) but the policy works perfectly fine (probably thanks to [Policy Discovery](https://laravel.com/docs/11.x/authorization#policy-discovery))

Thanks for this great plugin by the way. It works great!

**Comment by emmadesilva** [comment]
Just discovered this UI today, it's amazing!

I think a small dictionary for common PascalCase names could be added.

For example I have model called `App\Models\GitHubRepository`, but the UI suggests `git_hub_repository_id` when it should be `github_repository_id`

Adding common names like GitHub could a small but nice improvement.


I'm also missing an option to create a (Pest) test?

**Comment by krisalcordo** [comment]
Is there a way to add a relationship after the model is created or do we have to manually do this for now?

**Comment by spitfire64** [comment]
Maybe a checkbox for creating a simple Livewire component?
EDIT: Amazing stuff, thanks!

**Comment by bmckay959** [comment]
> Great plugin!
> 
> I deal with a bunch of 3rd party data where the table structure is essentially already done. A lot of these have a bunch of columns(50+ sometimes).
> 
> * It would be nice when 'tabbing' to create a new field, if the scroll area would auto scroll down to see the new input.

Hello, I just wanted to bring this back in the discussion. I would really love this feature. I deal with a large number of columns in tables. The first 15-20 columns are great and allow me to add faster than ever before, but once the fields come out of view and I tab to create a new field, it would be really nice if the list of fields would scroll to the active form field.

7 fields fit in the UI...once I try to create the 8th, you can see the field is only half visible and the 9th field is completely hidden and I have to move to my mouse to scroll down.

<img width="900" alt="Screenshot 2024-07-19 at 11 17 40 AM" src="https://github.com/user-attachments/assets/0a21354f-88f4-450a-aa9b-b1303fe47411">
<img width="898" alt="Screenshot 2024-07-19 at 11 18 41 AM" src="https://github.com/user-attachments/assets/3a752762-d8e0-4432-bf6f-46287b3ea609">

**Comment by adelf** [comment]
Hi, @bmckay959. Thank you for mentioning this again) and sorry for this being not fixed all this time. Fixed. I'll try to release an update soon.

**Comment by adelf** [comment]
I'm closing this issue, since this UI looks finished for me now, so, please create new issues.

For all: the main idea of this UI to be as lightweight as possible. I need field and their types to create corresponding code: migrations, factories, etc. I don't need default values and indices for that. You can add it in the migration, that's why the migration file opens first after creating a model.

@kjaonline, I'm going to add a special actions, like "Add Field to existing model", or "Add Relation to existing model". They will try to add fields and relations to model, and also some related classes: factory, resource, etc.

@caendesilva I made field names depending on table name before, so if you have `$table = "github_repositories"`, it will suggest `github_repository_id` field name.
I don't know which Pest test should be created. Usually tests are testing some Use Cases. Like publishing the post. Making tests for CRUD operations... maybe only some base template. If you have any ideas - please share.

@spitfire64, what kind of Livewire component should it be? It's hard to guess, but if you have any ideas - please, welcome.

@altayevrim I've already fixed the policy issue. About boolean field defaults... where should they be? Migrations? Or model $attributes field?

**Comment by krisalcordo** [comment]
@adelf would it help for me to create a new issue for this?

> @kjaonline, I'm going to add a special actions, like "Add Field to existing model", or "Add Relation to existing model". They will try to add fields and relations to model, and also some related classes: factory, resource, etc.

**Comment by adelf** [comment]
@kjaonline it's not necessary. It's already in our internal task tracker.

**Comment by spitfire64** [comment]
@adelf Just by creating a component named Index.php in a subfolder (given the model name) instead of an ordinary Controller would have been nice.  Inside this file, why not define all columns as properties, then maybe the mount() and render() methods empty. The render might return a view in a corresponding subolder named index.blade.php.

Another suggestion: Your template engine could also generate some lines that can be copied: The routes. (either traditional CRUD-rotes or a Livewire-route, depending on the checkbox named "livewire")
